src/goto-instrument/contracts/dynamic-frames/dfcc_check_loop_normal_form.cpp

```diff
@@ -0,0 +1,203 @@
+      {
+        // clang-format off
+        log.error() << "Loop spanning:"
+                    << "\n- head: " << head->source_location()
+                    << "\n- latch: " << latch->source_location()
+                    << "\noverlaps loop spanning:"
+                    << "\n- head: " << span.first->source_location()
+                    << "\n- latch: " << span.second->source_location()
+                    << messaget::eom;
+        // clang-format on
+        throw analysis_exceptiont(
+          "Found loops with overlapping instruction spans");
+      }
+    }
+
+    spans.push_back({head, latch});
+
+    // Check that:
+    // 1. all loop instructions are in the range [head, latch]
+    // 2. except for the head instruction, no incoming edge from outside the
+    // loop
+    for(const auto &i : loop_instructions)
+    {
+      if(
+        i->location_number < head->location_number ||
+        i->location_number > latch->location_number)
+      {
+        // clang-format off
+        log.error() << "Loop body instruction at:"
+                    << "\n- " << i->source_location()
+                    << "\nis outside of"
+                    << "\n- head: " << head->source_location()
+                    << "\n- latch: " << latch->source_location()
+                    << messaget::eom;
+        // clang-format on
+        throw analysis_exceptiont(
+          "Found loop body instruction outside of the [head, latch] "
+          "instruction span");
+      }
+
+      for(const auto &from : i->incoming_edges)
+      {
+        if(i != head && !loop_instructions.contains(from))
+        {
+          // clang-format off
+        log.error() << "Loop body instruction at:"
+                    << "\n- " << i->source_location()
+                    << "\n has incoming edge from outside the loop at:"
+                    << "\n- " << from->source_location()
+                    << messaget::eom;
+          // clang-format on
+
+          throw analysis_exceptiont(
+            "Found loop body instruction with incoming edge from outside the "
+            "loop");
+        }
+      }
+    }
+
+    // Check if a loop contains another loop's head (resp. latch) then
+    // it also contains the latch (resp. head)
+    for(auto &other_loop : natural_loops.loop_map)
+    {
+      auto &other_loop_instructions = other_loop.second;
+      bool contains_head = other_loop_instructions.contains(head);
+      bool contains_latch = other_loop_instructions.contains(latch);
+      if(contains_head != contains_latch)
+      {
+        log.error() << "Loop spanning from " << head->source_location()
+                    << " to " << latch->source_location()
+                    << "contains the head or latch of the loop spanning from "
+                    << other_loop.first->source_location() << messaget::eom;
+        throw analysis_exceptiont(
+          "Found two loops with overlapping instruction spans");
```

Comment:
We have already checked overlapping loops above on line 97. The check here is redundant.

Reply:
These are slightly different checks, check 1)  deals with inclusion of spans of instructions regardless of the loop nesting relation, check 2) deals with nesting of natural loop instruction sets and can be seen as checking an invariant the output of the natural_loops, for which we do not have a clear spec. So i'd like to keep both checks in place.